python/pyarrow/tests/test_compute_kernels.py

```diff
@@ -0,0 +1,1237 @@
+        return f'{self.function_name}({arg_str})'
+
+
+def get_sample_empty_calls():
+    for function_name in pc.function_registry().list_functions():
+        if function_name not in function_taxonomy:
+            continue
+        for signature in function_taxonomy[function_name]:
+            for arg_types, _return_type, parameters_map in get_sample_calls(
+                    signature):
+                args = [pa.array([], type=arg_type) for arg_type in arg_types]
+                options_fn = samples_opts.get(function_name, None)
+                options = None
+                if options_fn is not None:
+                    options = options_fn(arg_types)
+                yield SampleCall(function_name, args, options, parameters_map)
+
+
+@pytest.mark.parametrize('function_name',
+                         pc.function_registry().list_functions(), ids=str)
+def test_all_functions_in_taxonomy(function_name):
+    if function_name.startswith('array_'):
+        pytest.xfail(
+            'ARROW-13873: array_* functions should be hidden from python')
+    if function_name.endswith('_meta_binary'):
+        pytest.xfail(
+            'ARROW-13949: *_meta_binary functions should be '
+            'hidden from python')
+    if function_name.startswith('hash_'):
+        pytest.xfail(
+            'ARROW-13943: hash_* functions should be hidden from python')
+    if function_name == 'list_element':
+        pytest.xfail('ARROW-13594: list_element requires a scalar input')
+    assert function_name in function_taxonomy
+
+
+def _check_expect_fail(sample_call: SampleCall, message: str,
+                       signatures_map: List[Tuple[Set[str], List[str]]]):
+    for possible_sig in signatures_map:
+        if sample_call.function_name in possible_sig[0]:
+            signatures = possible_sig[1]
+            for signature in signatures:
+                parsed_signature = parse_function_signature(signature)
+                if parsed_signature.matches_parameters(
+                        sample_call.parameters_map):
+                    pytest.xfail(message)
+
+
+def safe_str(o):
+    raw_str = str(o)
+    return re.sub(r'\s', '', raw_str).replace('(', '-').replace(')', '-')
+
+
+@pytest.mark.parametrize('sample_call', get_sample_empty_calls(), ids=safe_str)
+def test_supports_empty_arrays(sample_call):
+    _check_expect_fail(sample_call,
+                       'ARROW-13390: Improve type support for coalesce kernel',
+                       [
+                           ({'choose'}, ['<T:list,I:integral>(I,T...)=>T',
+                                         '<T:struct,I:integral>(I,T...)=>T']),
+                           ({'coalesce'},
+                            ['<T:list>(T)=>T', '<T:struct>(T)=>T']),
+                           ({'replace_with_mask'}, [
+                               '<T:list,B:boolean>(T,B,T)=>T',
+                               '<T:struct,B:boolean>(T,B,T)=>T'])
+                       ])
```

Comment:
Wouldn't these be re-parsed and re-checked on every iteration of the test? That might inflate runtime…

Reply:
I tried caching at one point and it didn't have any real effect.